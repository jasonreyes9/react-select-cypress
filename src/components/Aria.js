// @flow
import React, { type Node } from 'react';

import { SROnly } from '../primitives';
import { className } from '../utils';

type StatusProps = { children: Node };
export const AriaStatus = (props: StatusProps) => (
  <SROnly className={className('aria-status')} {...props} />
);
