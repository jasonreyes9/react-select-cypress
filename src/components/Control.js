// @flow
import React from 'react';

import { className } from '../utils';
import { Div } from '../primitives';
import { borderRadius, colors, spacing } from '../theme';
import { type PropsWithStyles } from '../types';

type ControlProps = { isDisabled: boolean, isFocused: boolean };
type Props = PropsWithStyles & ControlProps;

export const css = ({ isDisabled, isFocused }: ControlProps) => ({
  alignItems: 'center',
  backgroundColor: isDisabled
    ? colors.neutral5
    : isFocused ? colors.neutral0 : colors.neutral2,
  borderColor: isDisabled
    ? colors.neutral10
    : isFocused ? colors.primary : colors.neutral20,
  borderRadius: borderRadius,
  borderStyle: 'solid',
  borderWidth: 1,
  boxShadow: isFocused ? `0 0 0 1px ${colors.primary}` : null,
  cursor: 'default',
  display: 'flex ',
  flexWrap: 'wrap',
  justifyContent: 'space-between',
  minHeight: spacing.controlHeight,
  outline: '0 !important',
  position: 'relative',
  transition: 'all 100ms',

  '&:hover': {
    borderColor: isFocused ? colors.primary : colors.neutral30,
  },
});

const Control = (props: Props) => {
  const { getStyles, isDisabled, isFocused, ...cleanProps } = props;
  return (
    <Div
      className={className('control', { isDisabled, isFocused })}
      css={getStyles('control', props)}
      {...cleanProps}
    />
  );
};

export default Control;
