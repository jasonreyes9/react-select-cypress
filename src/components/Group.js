// @flow
import React, { type Node, type ComponentType } from 'react';

import { Div } from '../primitives';
import { spacing } from '../theme';
import type { CommonProps } from '../types';

type ComponentProps = {
  /** The children to be rendered. */
  children: Node,
  /** Component to wrap the label, recieves headingProps. */
  Heading: ComponentType<any>,
  /** Props passed to the heading. */
  headingProps: {
    'aria-labelledby': string,
  },
  /** props passed to the wrapping element for the group. */
  innerProps: {
    'aria-expanded': boolean,
    'aria-label': string,
    role: 'group',
  },
  /** Label to be displayed in the heading component. */
  label: Node,
};
export type GroupProps = CommonProps & ComponentProps;

export const groupCSS = () => ({
  paddingBottom: spacing.baseUnit * 2,
  paddingTop: spacing.baseUnit * 2,
});

const Group = (props: GroupProps) => {
  const {
    children,
    cx,
    getStyles,
    Heading,
    headingProps,
    label,
    innerProps,
  } = props;
  return (
    <Div
      className={cx('group')}
      css={getStyles('group', props)}
      {...innerProps}
    >
      <Heading getStyles={getStyles} cx={cx} {...headingProps}>
        {label}
      </Heading>
      <Div>{children}</Div>
    </Div>
  );
};

export const groupHeadingCSS = () => ({
  color: '#999',
  cursor: 'default',
  display: 'block',
  fontSize: '75%',
  fontWeight: '500',
  marginBottom: '0.25em',
  paddingLeft: spacing.baseUnit * 3,
  paddingRight: spacing.baseUnit * 3,
  textTransform: 'uppercase',
});

export const GroupHeading = (props: any) => {
  const { cx, getStyles, ...cleanProps } = props;
  return (
    <Div
      className={cx('group-heading')}
      css={getStyles('groupHeading', props)}
      {...cleanProps}
    />
  );
};

export default Group;
