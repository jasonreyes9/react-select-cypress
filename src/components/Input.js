// @flow
import React, { type ElementRef } from 'react';
import AutosizeInput from 'react-input-autosize';

import { className } from '../utils';
import { spacing } from '../theme';
import { Div } from '../primitives';

import type { PropsWithStyles } from '../types';

export type InputProps = PropsWithStyles & {
  innerRef: (ElementRef<*>) => void,
  isHidden: boolean,
};

export const css = () => ({
  margin: spacing.baseUnit / 2,
  paddingBottom: spacing.baseUnit / 2,
  paddingTop: spacing.baseUnit / 2,
});
const inputStyle = isHidden => ({
  background: 0,
  border: 0,
  fontSize: 'inherit',
  opacity: isHidden ? 0 : 1,
  outline: 0,
  padding: 0,
});

const Input = ({ getStyles, innerRef, isHidden, ...props }: InputProps) => (
  <Div css={getStyles('input', props)}>
    <AutosizeInput
      className={className('input')}
      inputRef={innerRef}
      inputStyle={inputStyle(isHidden)}
      {...props}
    />
  </Div>
);
export default Input;
