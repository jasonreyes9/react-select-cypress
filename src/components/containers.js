// @flow
import React, { Component } from 'react';

import { className } from '../utils';
import { Div } from '../primitives';
import { paddingHorizontal, paddingVertical } from '../mixins';
import { spacing } from '../theme';
import { type PropsWithStyles } from '../types';

// ==============================
// Root Container
// ==============================

type ContainerProps = PropsWithStyles & { isDisabled: boolean };
export const containerCSS = ({ isDisabled }: { isDisabled: boolean }) => ({
  pointerEvents: isDisabled ? 'none' : 'initial', // cancel mouse events when disabled
  position: 'relative',
});
export const SelectContainer = ({ getStyles, ...props }: ContainerProps) => {
  const { isDisabled, ...cleanProps } = props;
  return (
    <Div
      css={getStyles('container', props)}
      className={className('container', { isDisabled })}
      {...cleanProps}
    />
  );
};

// ==============================
// Value Container
// ==============================

type ValueContainerProps = PropsWithStyles & {
  isMulti: boolean,
  hasValue: boolean,
  maxHeight: number,
};
export const valueContainerCSS = ({ maxHeight }: ValueContainerProps) => ({
  alignItems: 'baseline',
  display: 'flex ',
  flex: 1,
  flexWrap: 'wrap',
  maxHeight: maxHeight, // max-height allows scroll when multi
  overflowY: 'auto',
  ...paddingHorizontal(spacing.baseUnit * 2),
  ...paddingVertical(spacing.baseUnit / 2),
});
export class ValueContainer extends Component<ValueContainerProps> {
  shouldScrollBottom: boolean = false;
  node: HTMLElement;
  componentWillUpdate() {
    if (!this.props.isMulti) return;

    // scroll only if the user was already at the bottom
    const total = this.node.scrollTop + this.node.offsetHeight;
    this.shouldScrollBottom = total === this.node.scrollHeight;
  }
  componentDidUpdate() {
    if (!this.props.isMulti) return;

    // ensure we're showing items being added by forcing scroll to the bottom
    if (this.shouldScrollBottom) {
      this.node.scrollTop = this.node.scrollHeight;
    }
  }
  getScrollContainer = (ref: HTMLElement) => {
    this.node = ref;
  };
  render() {
    const {
      isMulti,
      getStyles,
      hasValue,
      maxHeight, // Unused var: invalid DOM attribute, React will warn if not removed
      ...cleanProps
    } = this.props;

    return (
      <Div
        innerRef={isMulti ? this.getScrollContainer : null}
        className={className('value-container', { isMulti, hasValue })}
        css={getStyles('valueContainer', this.props)}
        {...cleanProps}
      />
    );
  }
}

// ==============================
// Indicator Container
// ==============================

export const indicatorsContainerCSS = () => ({
  display: 'flex ',
  flexShrink: 0,
});
export const IndicatorsContainer = ({
  getStyles,
  ...props
}: PropsWithStyles) => {
  return (
    <Div
      className={className('indicators')}
      css={getStyles('indicatorsContainer', props)}
      {...props}
    />
  );
};
