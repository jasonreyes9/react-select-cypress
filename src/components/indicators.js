// @flow
// @jsx glam
import React, { type ElementType } from 'react';
import glam from 'glam';

import { className } from '../utils';
import { Div, Span, SROnly } from '../primitives';
import { colors, spacing } from '../theme';
import { type PropsWithStyles } from '../types';

// ==============================
// Dropdown & Clear Icons
// ==============================

const Svg = ({ size, ...props }: { size: number }) => (
  <svg
    height={size}
    width={size}
    viewBox="0 0 20 20"
    css={{
      display: 'inline-block',
      fill: 'currentColor',
      lineHeight: 1,
      stroke: 'currentColor',
      strokeWidth: 0,
    }}
    {...props}
  />
);
export const CrossIcon = (props: any) => (
  <Svg
    size={20}
    focusable="false"
    role="presentation"
    className={className(['icon', 'cross-icon'])}
    {...props}
  >
    <path d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z" />
  </Svg>
);
export const DownChevron = (props: any) => (
  <Svg
    size={20}
    focusable="false"
    role="presentation"
    className={className(['icon', 'down-icon'])}
    {...props}
  >
    <path d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z" />
  </Svg>
);

// ==============================
// Dropdown & Clear Buttons
// ==============================

type IndicatorProps = PropsWithStyles & {
  children: ElementType,
  isFocused: boolean,
  innerProps: any,
};

export const css = ({ isFocused }: IndicatorProps) => ({
  color: isFocused ? colors.text : colors.neutral20,
  cursor: 'pointer',
  display: 'flex ',
  padding: '8px 2px',
  transition: 'opacity 200ms',

  ':first-child': { paddingLeft: spacing.baseUnit * 2 },
  ':last-child': { paddingRight: spacing.baseUnit * 2 },

  ':hover': {
    color: isFocused ? colors.text : colors.neutral40,
  },
});

export const DropdownIndicator = (props: IndicatorProps) => {
  const { children, getStyles, innerProps } = props;
  return (
    <Div
      {...innerProps}
      css={getStyles('indicator', props)}
      className={className(['indicator', 'dropdown-indicator'])}
    >
      {children}
    </Div>
  );
};
DropdownIndicator.defaultProps = {
  children: <DownChevron />,
};

export const ClearIndicator = (props: IndicatorProps) => {
  const { children, getStyles, innerProps } = props;
  return (
    <Div
      {...innerProps}
      css={getStyles('indicator', props)}
      className={className(['indicator', 'clear-indicator'])}
    >
      {children}
    </Div>
  );
};
ClearIndicator.defaultProps = {
  children: <CrossIcon />,
};

// ==============================
// Loading
// ==============================

const keyframesName = 'react-select-loading-indicator';

const LoadingContainer = ({ size }: { size: number }) => (
  <Div
    css={{
      alignSelf: 'center',
      fontSize: size,
      lineHeight: 1,
      marginRight: size,
      textAlign: 'center',
      verticalAlign: 'middle',
    }}
  />
);
type DotProps = { color: string, delay: number, offset: boolean };
const LoadingDot = ({ color, delay, offset }: DotProps) => (
  <Span
    css={{
      animationDuration: '1s',
      animationDelay: `${delay}ms`,
      animationIterationCount: 'infinite',
      animationName: keyframesName,
      animationTimingFunction: 'ease-in-out',
      backgroundColor: color,
      borderRadius: '1em',
      display: 'inline-block',
      marginLeft: offset ? '1em' : null,
      height: '1em',
      verticalAlign: 'top',
      width: '1em',
    }}
  />
);
// TODO @jossmac Source `keyframes` solution for glam
// - at the very least, ensure this is only rendered once to the DOM
const LoadingAnimation = () => (
  <style type="text/css">
    {`@keyframes ${keyframesName} {
        0%, 80%, 100% { opacity: 0; }
        40% { opacity: 1; }
    };`}
  </style>
);

type LoadingIconProps = { isFocused: boolean, size: number };
const LoadingIcon = ({ isFocused, size = 4 }: LoadingIconProps) => {
  const clr = isFocused ? colors.text : colors.neutral20;

  return (
    <LoadingContainer size={size}>
      <LoadingAnimation />
      <LoadingDot color={clr} />
      <LoadingDot color={clr} delay={160} offset />
      <LoadingDot color={clr} delay={320} offset />
      <SROnly>Loading</SROnly>
    </LoadingContainer>
  );
};

export const LoadingIndicator = (props: IndicatorProps) => {
  const { getStyles, innerProps } = props;
  return (
    <Div
      {...innerProps}
      css={getStyles('indicator', props)}
      className={className(['indicator', 'loading-indicator'])}
    >
      <LoadingIcon isFocused={props.isFocused} />
    </Div>
  );
};
