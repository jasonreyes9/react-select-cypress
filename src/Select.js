import React from 'react';
import ReactDOM from 'react-dom';
import Input from 'react-input-autosize';
import classNames from 'classnames';

import MultiValue from './MultiValue';
import Option from './Option';
import SingleValue from './SingleValue';

var requestId = 0;

var Select = React.createClass({

	displayName: 'Select',

	propTypes: {
		addLabelText: React.PropTypes.string,       // placeholder displayed when you want to add a label on a multi-value input
		allowCreate: React.PropTypes.bool,          // whether to allow creation of new entries
		asyncOptions: React.PropTypes.func,         // function to call to get options
		autoload: React.PropTypes.bool,             // whether to auto-load the default async options set
		backspaceRemoves: React.PropTypes.bool,     // whether backspace removes an item if there is no text input
		cacheAsyncResults: React.PropTypes.bool,    // whether to allow cache
		className: React.PropTypes.string,          // className for the outer element
		clearAllText: React.PropTypes.string,       // title for the "clear" control when multi: true
		clearValueText: React.PropTypes.string,     // title for the "clear" control
		clearable: React.PropTypes.bool,            // should it be possible to reset value
		delimiter: React.PropTypes.string,          // delimiter to use to join multiple values
		disabled: React.PropTypes.bool,             // whether the Select is disabled or not
		filterOption: React.PropTypes.func,         // method to filter a single option  (option, filterString)
		filterOptions: React.PropTypes.func,        // method to filter the options array: function ([options], filterString, [values])
		ignoreCase: React.PropTypes.bool,           // whether to perform case-insensitive filtering
		inputProps: React.PropTypes.object,         // custom attributes for the Input (in the Select-control) e.g: {'data-foo': 'bar'}
		isLoading: React.PropTypes.bool,            // whether the Select is loading externally or not (such as options being loaded)
		labelKey: React.PropTypes.string,           // path of the label value in option objects
		matchPos: React.PropTypes.string,           // (any|start) match the start or entire string when filtering
		matchProp: React.PropTypes.string,          // (any|label|value) which option property to filter on
		multi: React.PropTypes.bool,                // multi-value input
		multiValueComponent: React.PropTypes.func,  // value component to render when multiple is set to true
		name: React.PropTypes.string,               // field name, for hidden <input /> tag
		newOptionCreator: React.PropTypes.func,     // factory to create new options when allowCreate set
		noResultsText: React.PropTypes.string,      // placeholder displayed when there are no matching search results
		onBlur: React.PropTypes.func,               // onBlur handler: function (event) {}
		onChange: React.PropTypes.func,             // onChange handler: function (newValue) {}
		onFocus: React.PropTypes.func,              // onFocus handler: function (event) {}
		onInputChange: React.PropTypes.func,        // onInputChange handler: function (inputValue) {}
		onOptionLabelClick: React.PropTypes.func,   // onCLick handler for value labels: function (value, event) {}
		optionComponent: React.PropTypes.func,      // option component to render in dropdown
		optionRenderer: React.PropTypes.func,       // optionRenderer: function (option) {}
		options: React.PropTypes.array,             // array of options
		placeholder: React.PropTypes.string,        // field placeholder, displayed when there's no value
		searchable: React.PropTypes.bool,           // whether to enable searching feature or not
		searchingText: React.PropTypes.string,      // message to display whilst options are loading via asyncOptions
		searchPromptText: React.PropTypes.string,   // label to prompt for search input
		singleValueComponent: React.PropTypes.func, // value component to render when multiple is set to false
		value: React.PropTypes.any,                 // initial field value
		valueKey: React.PropTypes.string,           // path of the label value in option objects
		valueRenderer: React.PropTypes.func         // valueRenderer: function (option) {}
	},

	getDefaultProps () {
		return {
			addLabelText: 'Add "{label}"?',
			allowCreate: false,
			autoload: true,
			backspaceRemoves: true,
			cacheAsyncResults: true,
			clearAllText: 'Clear all',
			clearValueText: 'Clear value',
			clearable: true,
			delimiter: ',',
			disabled: false,
			ignoreCase: true,
			inputProps: {},
			isLoading: false,
			labelKey: 'label',
			matchPos: 'any',
			matchProp: 'any',
			multi: false,
			multiValueComponent: MultiValue,
			noResultsText: 'No results found',
			optionComponent: Option,
			placeholder: 'Select...',
			searchable: true,
			searchingText: 'Searching...',
			searchPromptText: 'Type to search',
			singleValueComponent: SingleValue,
			valueKey: 'value',
		};
	},

	getInitialState () {
		return {
			isFocused: false,
			isLoading: false,
			isOpen: false
		};
	},

	focus () {
		if (!this.refs.input) return;
		this.refs.input.focus();
	},

	getOptionLabel (op) {
		return op[this.props.labelKey];
	},

	handleMouseDown (event) {
		// if the event was triggered by a mousedown and not the primary
		// button, or if the component is disabled, ignore it.
		if (this.props.disabled || (event.type === 'mousedown' && event.button !== 0)) {
			return;
		}

		// prevent default event handlers
		event.stopPropagation();
		event.preventDefault();

		// for the non-searchable select, close the dropdown when button is clicked
		if (this.state.isOpen && !this.props.searchable) {
			this.setState({
				isOpen: false
			});
			return;
		}

		if (this.state.isFocused) {
			// if the input is focused, ensure the menu is open
			this.setState({
				isOpen: true
			});
		} else {
			// otherwise, focus the input and open the menu
			this._openAfterFocus = true;
			this.focus();
		}
	},

	handleMouseDownOnArrow (event) {
		// if the event was triggered by a mousedown and not the primary
		// button, or if the component is disabled, ignore it.
		if (this.props.disabled || (event.type === 'mousedown' && event.button !== 0)) {
			return;
		}
		// If the menu isn't open, let the event bubble to the main handleMouseDown
		if (!this.state.isOpen) {
			return;
		}
		// prevent default event handlers
		event.stopPropagation();
		event.preventDefault();
		// close the menu
		this.setState({
			isOpen: false
		});
	},

	handleInputFocus (event) {
		var isOpen = this.state.isOpen || this._openAfterFocus;
		if (this.props.onFocus) {
			this.props.onFocus(event);
		}
		this.setState({
			isFocused: true,
			isOpen: isOpen
		});
		this._openAfterFocus = false;
	},

	handleInputBlur (event) {
		if (document.activeElement.isEqualNode(this.refs.menu)) {
			return;
		}
		if (this.props.onBlur) {
			this.props.onBlur(event);
		}
		this.setState({
			inputValue: '',
			isFocused: false,
			isOpen: false
		});
	},

	handleInputChange (event) {
		if (this.props.onInputChange) {
			this.props.onInputChange(event.target.value);
		}
		if (this.props.asyncOptions) {
			this.setState({
				// isLoading: true,
				inputValue: event.target.value
			});
			// this.loadAsyncOptions(event.target.value, {
			// 	isLoading: false,
			// 	isOpen: true
			// }, this._bindCloseMenuIfClickedOutside);
		} else {
			// var filteredOptions = this.filterOptions(this.state.options);
			this.setState({
				isOpen: true,
				inputValue: event.target.value,
				// filteredOptions: filteredOptions,
				// focusedOption: this._getNewFocusedOption(filteredOptions)
			});
		}
	},

	getValueArray () {
		if (this.props.multi) {
			if (!Array.isArray(this.props.value)) return [];
			return this.props.value.map(this.expandValue).filter(i => i);
		}
		var expandedValue = this.expandValue(this.props.value);
		return expandedValue ? [expandedValue] : [];
	},

	expandValue (value) {
		if (typeof value !== 'string') return value;
		if (!this.props.options) return;
		for (var i = 0; i < this.props.options.length; i++) {
			if (this.props.options[i][this.props.valueKey] === value) return this.props.options[i];
		}
	},

	setValue (value) {
		if (this.props.onChange) {
			this.props.onChange(value);
		}
	},

	selectValue (value) {
		if (!this.props.multi) {
			this.setValue(value);
		} else if (value) {
			this.addValue(value);
		}
	},

	addValue (value) {
		var valueArray = this.getValueArray();
		this.setValue(valueArray.concat(value));
	},

	popValue () {
		var valueArray = this.getValueArray();
		if (!valueArray.length) return;
		this.setValue(valueArray.slice(0, valueArray.length - 1));
	},

	removeValue (value) {
		var valueArray = this.getValueArray();
		this.setValue(valueArray.filter(i => i !== value));
	},

	clearValue (event) {
		// if the event was triggered by a mousedown and not the primary
		// button, ignore it.
		if (event && event.type === 'mousedown' && event.button !== 0) {
			return;
		}
		event.stopPropagation();
		event.preventDefault();
		this.setValue(null);
	},

	focusOption (option) {
		this.setState({
			focusedOption: option
		});
	},

	unfocusOption (option) {
		if (this.state.focusedOption === option) {
			this.setState({
				focusedOption: null
			});
		}
	},

	isLoading () {
		return this.props.isLoading || this.state.isLoading;
	},

	renderLoading () {
		if (!this.isLoading()) return;
		return (
			<span className="Select-loading-zone" aria-hidden="true">
				<span className="Select-loading" />
			</span>
		);
	},

	renderValue (valueArray) {
		var renderLabel = this.props.valueRenderer || this.getOptionLabel;
		var MultiValueComponent = this.props.multiValueComponent;
		var SingleValueComponent = this.props.singleValueComponent;
		if (!valueArray.length) {
			return <div className="Select-placeholder">{this.props.placeholder}</div>;
		}
		if (this.props.multi) {
			return valueArray.map(i => {
				return (
					<MultiValueComponent
						key={i[this.props.valueKey]}
						value={i}
						onRemove={this.removeValue}
						disabled={this.props.disabled}
						>
						{renderLabel(i)}
					</MultiValueComponent>
				);
			});
		} else if (!this.state.inputValue) {
			return (
				<SingleValueComponent
					value={valueArray[0]}>
					{renderLabel(valueArray[0])}
				</SingleValueComponent>
			);
		}
	},

	renderInput (valueArray) {
		var className = classNames('Select-input', this.props.inputProps.className);
		if (this.props.disabled || !this.props.searchable) {
			if (this.props.multi && valueArray.length) return;
			return <div className={className}>&nbsp;</div>;
		}
		return (
			<Input
				{...this.props.inputProps}
				className={className}
				tabIndex={this.props.tabIndex}
				onBlur={this.handleInputBlur}
				onChange={this.handleInputChange}
				onFocus={this.handleInputFocus}
				minWidth="5"
				ref="input"
				value={this.state.inputValue}
			/>
		);
	},

	renderClear () {
		if (!this.props.clearable || !this.props.value || (this.props.multi && !this.props.value.length) || this.props.disabled || this.isLoading()) return;
		return (
			<span className="Select-clear-zone" title={this.props.multi ? this.props.clearAllText : this.props.clearValueText} aria-label={this.props.multi ? this.props.clearAllText : this.props.clearValueText} onMouseDown={this.clearValue} onTouchEnd={this.clearValue} onClick={this.clearValue}>
				<span className="Select-clear" dangerouslySetInnerHTML={{ __html: '&times;' }} />
			</span>
		);
	},

	renderArrow () {
		return (
			<span className="Select-arrow-zone" onMouseDown={this.handleMouseDownOnArrow}>
				<span className="Select-arrow" onMouseDown={this.handleMouseDownOnArrow} />
			</span>
		);
	},

	renderMenuOptions (valueArray) {
		// TODO: filter options
		// TODO: don't re-gen unless options / input have changed
		// TODO: other ways of getting options (async options)...
		if (this.props.options && this.props.options.length) {
			var Option = this.props.optionComponent;
			var renderLabel = this.props.optionRenderer || this.getOptionLabel;
			return this.props.options.map(option => {
				var isSelected = valueArray.indexOf(option) > -1;
				var isFocused = option === this.state.focusedOption;
				var optionRef = isFocused ? 'focused' : null;
				var optionClass = classNames({
					'Select-option': true,
					'is-selected': isSelected,
					'is-focused': isFocused,
					'is-disabled': option.disabled,
				});
				return (
					<Option
						className={optionClass}
						key={'option-' + option[this.props.valueKey]}
						mouseDown={this.selectValue}
						mouseEnter={this.focusOption}
						mouseLeave={this.unfocusOption}
						option={option}
						ref={optionRef}
						>
						{renderLabel(option)}
					</Option>
				);
			});
		} else {
			var noResultsText;
			var noResultsClass;
			if (this.isLoading()) {
				noResultsClass = 'Select-searching';
				noResultsText = this.props.searchingText;
			} else if (this.state.inputValue || !this.props.asyncOptions) {
				noResultsClass = 'Select-noresults';
				noResultsText = this.props.noResultsText;
			} else {
				noResultsClass = 'Select-search-prompt';
				noResultsText = this.props.searchPromptText;
			}
			return (
				<div className={noResultsClass}>
					{noResultsText}
				</div>
			);
		}
	},

	renderMenu (valueArray) {
		if (!this.state.isOpen) return;
		return (
			<div ref="menuContainer" className="Select-menu-outer">
				<div ref="menu" className="Select-menu" onMouseDown={this.handleMouseDownOnMenu}>
					{this.renderMenuOptions(valueArray)}
				</div>
			</div>
		);
	},

	getWrapperClassName (hasValue) {
		return classNames('Select', this.props.className, {
			'Select--multi': this.props.multi,
			'is-searchable': this.props.searchable,
			'is-open': this.state.isOpen,
			'is-focused': this.state.isFocused,
			'is-loading': this.isLoading(),
			'is-disabled': this.props.disabled,
			'has-value': hasValue,
		});
	},

	render () {
		let valueArray = this.getValueArray();
		return (
			<div ref="wrapper" className={this.getWrapperClassName(valueArray.length)}>
				<input type="hidden" ref="value" name={this.props.name} value={this.state.value} disabled={this.props.disabled} />
				<div className="Select-control" ref="control" onKeyDown={this.handleKeyDown} onMouseDown={this.handleMouseDown} onTouchEnd={this.handleMouseDown}>
					{this.renderValue(valueArray)}
					{this.renderInput(valueArray)}
					{this.renderLoading()}
					{this.renderClear()}
					{this.renderArrow()}
				</div>
				{this.renderMenu(valueArray)}
			</div>
		);
	}

});

export default Select;
