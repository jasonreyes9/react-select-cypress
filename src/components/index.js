// @flow
import {
  IndicatorsContainer,
  SelectContainer,
  ValueContainer,
} from './containers';
import {
  ClearIndicator,
  DropdownIndicator,
  LoadingIndicator,
  IndicatorSeparator,
} from './indicators';

import Control from './Control';
import Group, { GroupHeading } from './Group';
import Input from './Input';
import Menu, { MenuList, NoOptionsMessage, LoadingMessage } from './Menu';
import MultiValue, {
  MultiValueContainer,
  MultiValueLabel,
  MultiValueRemove,
} from './MultiValue';
import Option from './Option';
import Placeholder from './Placeholder';
import SingleValue from './SingleValue';

export type SelectComponents = {
  ClearIndicator: typeof ClearIndicator,
  Control: typeof Control,
  DropdownIndicator: typeof DropdownIndicator,
  Group: typeof Group,
  GroupHeading: typeof GroupHeading,
  IndicatorsContainer: typeof IndicatorsContainer,
  IndicatorSeparator: typeof IndicatorSeparator,
  Input: typeof Input,
  LoadingIndicator: typeof LoadingIndicator,
  Menu: typeof Menu,
  MenuList: typeof MenuList,
  LoadingMessage: typeof LoadingMessage,
  NoOptionsMessage: typeof NoOptionsMessage,
  MultiValue: typeof MultiValue,
  MultiValueContainer: typeof MultiValueContainer,
  MultiValueLabel: typeof MultiValueLabel,
  MultiValueRemove: typeof MultiValueRemove,
  Option: typeof Option,
  Placeholder: typeof Placeholder,
  SelectContainer: typeof SelectContainer,
  SingleValue: typeof SingleValue,
  ValueContainer: typeof ValueContainer,
};

export type SelectComponentsConfig = $Shape<SelectComponents>;

export const components: SelectComponents = {
  ClearIndicator: ClearIndicator,
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Group: Group,
  GroupHeading: GroupHeading,
  IndicatorsContainer: IndicatorsContainer,
  IndicatorSeparator: IndicatorSeparator,
  Input: Input,
  LoadingIndicator: LoadingIndicator,
  Menu: Menu,
  MenuList: MenuList,
  LoadingMessage: LoadingMessage,
  NoOptionsMessage: NoOptionsMessage,
  MultiValue: MultiValue,
  MultiValueContainer: MultiValueContainer,
  MultiValueLabel: MultiValueLabel,
  MultiValueRemove: MultiValueRemove,
  Option: Option,
  Placeholder: Placeholder,
  SelectContainer: SelectContainer,
  SingleValue: SingleValue,
  ValueContainer: ValueContainer,
};

type Props = {
  components: SelectComponentsConfig,
};

export const defaultComponents = (props: Props) => ({
  ...components,
  ...props.components,
});
