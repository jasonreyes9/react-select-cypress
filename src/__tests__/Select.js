import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import Select from '../Select';
import { components } from '../components';

const { Option, SingleValue } = components;

Enzyme.configure({ adapter: new Adapter() });

test('defaults', () => {
  const tree = shallow(<Select />);
  expect(toJson(tree)).toMatchSnapshot();
});

test('formatOptionLabel', () => {
  function formatOptionLabel({ label, value }, { context }) {
    return `${label} ${value} ${context}`;
  }
  const opts = [
    { label: '1', value: 'one' },
    { label: '2', value: 'two' },
    { label: '3', value: 'three' },
  ];
  const tree = shallow(
    <Select
      formatOptionLabel={formatOptionLabel}
      options={opts}
      value={opts[1]}
    />
  );
  const value = tree.find(SingleValue).at(0);
  expect(value.props().children).toBe('2 two value');
  tree.setProps({ menuIsOpen: true });
  const menu = tree.find(Option);
  expect(menu).toHaveLength(3);
  expect(menu.at(0).props().children).toBe('1 one menu');
  expect(menu.at(1).props().children).toBe('2 two menu');
  expect(menu.at(2).props().children).toBe('3 three menu');
});
