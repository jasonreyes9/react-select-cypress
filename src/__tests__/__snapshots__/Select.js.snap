// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
<SelectContainer
  getStyles={[Function]}
  isDisabled={false}
  onKeyDown={[Function]}
>
  <SROnly
    aria-atomic="true"
    aria-live="polite"
    role="status"
  >
    0
     results are available.
  </SROnly>
  <Control
    getStyles={[Function]}
    innerRef={[Function]}
    isDisabled={false}
    isFocused={false}
    onMouseDown={[Function]}
  >
    <ValueContainer
      getStyles={[Function]}
      hasValue={false}
      isMulti={false}
      maxHeight={100}
    >
      <Placeholder
        getStyles={[Function]}
        isDisabled={false}
        isMulti={false}
        key="placeholder"
      >
        Select...
      </Placeholder>
      <Input
        aria-autocomplete="list"
        aria-busy={false}
        aria-expanded={false}
        aria-haspopup={false}
        autoCapitalize="none"
        autoComplete="off"
        autoCorrect="off"
        getStyles={[Function]}
        id="react-select-2--input"
        innerRef={[Function]}
        isHidden={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        role="combobox"
        spellCheck="false"
        tabIndex="0"
        type="text"
        value=""
      />
    </ValueContainer>
    <IndicatorsContainer
      getStyles={[Function]}
    >
      <DropdownIndicator
        getStyles={[Function]}
        isFocused={false}
        onMouseDown={[Function]}
        role="button"
      >
        <DownChevron
          label="Toggle Menu"
        />
      </DropdownIndicator>
    </IndicatorsContainer>
  </Control>
</SelectContainer>
`;
